File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0388_Longest_Absolute_File_Path.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0388_Longest_Absolute_File_Path {
    public int lengthLongestPath(String input) {
        String[] paths = input.split("\n");
        int[] stack = new int[paths.length + 1];
        int res = 0, level, cur;
        for(String s : paths){
            // 层数
            level = s.lastIndexOf("\t") + 1;
            // 减去level个\t 增加1个/
            cur = stack[level + 1] = stack[level] + s.length() - level + 1;
            if(s.contains(".")) {
                // - 1 是去掉开始的/
                res = Math.max(res, cur - 1);
            }
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0388_Longest_Absolute_File_Path.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  int lengthLongestPath(string input) {
    vector<int> vec(1, 0);
    int res = 0, level = 0, cur = 0;
    bool isFile = false;
    for (int i = 0; i < input.size(); i++) {
      switch (input[i]) {
        case '\n':
          level = 0; cur = 0; isFile = false;
          break;
        case '\t':
          level++;
          break;
        case '.' :
          isFile = true;
        default:
          cur++;
          if (level + 1 >= vec.size()) {
            vec.push_back(0);
          }
          vec[level + 1] = vec[level] + cur + 1;
          if (isFile) {
            res = max(res, vec[level + 1] - 1);
          }
        }
    }
    return res;
  }
};
